Added a worklet implementation of the noisegate.

samhincks opened the pull request:
This PR has a single file implementing a worklet counterpart to audio-worklet/src/noisegate-script-processor.js. It resolves issue #102.

rtoy commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,194 @@
+      defaultValue: 0.0025,
+      minValue: 0,
+      maxValue: 1
+    }];
```
Add comments on how you picked these default values and the ranges.

rtoy commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,194 @@
+    this.previousWeight_ = 1.0;
+    this.envelope_ = new Float32Array(128);
+    this.weights_ = new Float32Array(128);
+    this.sampleRate = 44100;
```
Why 44100?

And why not trailing underscore?

hoch replied:
Because the sample rate is provided by the global scope.

samhincks replied:
So I should leave it as it is? This is sort of an edge case. I'm hardcoding the sample rate to be 44,100 but I want to match a variable in the global scope. I can file an issue to change this once getContextInfo() is implemented.

rtoy replied:
Add a TODO comment here to say it should use getContextInfo() to get the sample rate.

rtoy replied:
If you filed an issue on this, replace "TODO" with "TODO(#n)" where n is the issue number.  Or "TODO(url-to-issue)".

If you haven't filed an issue, file it, and do the above.

rtoy commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,194 @@
+    // envelope and its delay, with a higher value giving more smoothness at
+    // the expense of delay and vice versa.
+    this.alpha_ = this.getAlphaFromTimeConstant_(
+        parameters.timeConstant[0], 44100);
```
You should at least use `this.sampleRate` instead of 44100.

hoch replied:
`AudioWorkletProcessor` has `getContextInfo()` but it's not available in the custom build.

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,194 @@
+    this.alpha_ = this.getAlphaFromTimeConstant_(
+        parameters.timeConstant[0], 44100);
+
+    this.attack = parameters.attack[0];
```
Please add a comment; these are 'k-rate' params if you use them this way.

samhincks replied:
I should say that the a-rate audio params are used as k-rate audio params since the current build does not support k-rate audio params?

rtoy replied:
Or, maybe you can just actually make them a-rate?  But not in this PR.

rtoy replied:
This should also be mentioned in the JSDoc for this node.

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,194 @@
+      // a^2 / 2. Sine waves are therefore scaled back to the original
+      // amplitude, but other waveforms or constant sources can only be
+      // approximated.
+      const scaledEnvelopeValue
```
I would pull the local variables here outside of `for` loop. They alway are overwritten for each iteration, so no need to recreate the variables every time. (i.e. less garbage)

samhincks replied:
Sure... I'm just curios. Are we more concerned about memory efficiency in the worklet version as opposed to the script processor version?

rtoy replied:
Yes.  In the script processor if you generate garbage, you'll end up messing up the main thread and possibly the audio thread if GC takes too long.

In a worklet, if you generate garbage, you're pretty much guaranteed to glitch audio because GC on Chrome can take 10ms or more.

samhincks commented:
In the latest [commit](https://github.com/GoogleChrome/web-audio-samples/pull/107/commits/30e030378289cc62db168a192dec149a8a92d708), 
- The default values and ranges for audio params have been changed. And I added comments for these.
- Other minor changes based on review.

rtoy commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -148,22 +166,24 @@ registerProcessor('noisegate-audio-worklet',
+    // amplitude, but other waveforms or constant sources can only be
+    // approximated.
+    const scaledEnvelopeValue;
+    const weight;
```
How does this work?  These are const, but aren't initialized. And since they're const, they can't be changed.

samhincks commented:
In the latest [commit](https://github.com/GoogleChrome/web-audio-samples/pull/107/commits/c8ad8d4412a9088fb244121e52504226f1e65652), 
- Changed const to let (the code was submitted with a syntax error previously)
- Added a TODO comment.

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,216 @@
+    // to enable smooth transitions between sound and silence.
+    // The default value of 50ms has been set experimentally to minimize
+    // glitches in the demo.
+    {
```
Have you applied clang-format on this?

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,216 @@
+      minValue: 0,
+      maxValue: 0.1
+    },
+    // The maximum threshold is 0 since that is the defined maximum in dBFS.
```
I still find the comments are too verbose, but it's up to you.

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,216 @@
+    this.alpha_ = this.getAlphaFromTimeConstant_(
+        parameters.timeConstant[0], this.sampleRate);
+
+    // The a-rate audio-params are used as k-rate audio-params.
```
Perhaps

`// k-rate: use the first element of parameter data to process each render quantum.`

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,216 @@
+    let attackLossPerStep = 1;
+    let releaseGainPerStep = 1;
+
+    // TODO: Replace this weights-based approach for enabling attack/release
```
I think you should move this comment to l.149.

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,216 @@
+      releaseSteps = Math.ceil(this.sampleRate * this.release);
+      releaseGainPerStep = 1 / releaseSteps;
+    }
+    // For sine waves, the envelope eventually reaches an average power of
```
Add one line break at l.171. Also not sure if this comment is at the right place. It explains what's happening at l.182. No?

hoch commented on audio-worklet/src/noisegate-audio-worklet.js:
```diff
@@ -0,0 +1,216 @@
+    let scaledEnvelopeValue;
+    let weight;
+
+    // Compute an array of weights between 0 and 1 which will be multiplied with
```
The comment should explain "why" not about "what". This comments describe what is happening below.
